front/lib/api/llm/clients/openai/utils/openai_to_events.ts

```diff
@@ -0,0 +1,167 @@
+import type {
+  ResponseOutputItem,
+  ResponseOutputItemDoneEvent,
+  ResponseStreamEvent,
+} from "openai/resources/responses/responses";
+import type { Response } from "openai/resources/responses/responses";
+
+import type {
+  LLMOutputItem,
+  ProviderMetadata,
+} from "@app/lib/api/llm/types/events";
+import type { LLMEvent } from "@app/lib/api/llm/types/events";
+
+export async function* streamLLMEvents({
+  responseStreamEvents,
+  metadata,
+}: {
+  responseStreamEvents: AsyncIterable<ResponseStreamEvent>;
+  metadata: ProviderMetadata;
+}) {
+  for await (const event of responseStreamEvents) {
+    yield* toEvents({
+      event,
+      metadata,
+    });
+  }
+}
+
+function textDelta(delta: string, metadata: ProviderMetadata): LLMEvent {
+  return {
+    type: "text_delta",
+    content: {
+      delta,
+    },
+    metadata,
+  };
+}
+
+function reasoningDelta(delta: string, metadata: ProviderMetadata): LLMEvent {
+  return {
+    type: "reasoning_delta",
+    content: {
+      delta,
+    },
+    metadata,
+  };
+}
+
+function toolCall(
+  event: ResponseOutputItemDoneEvent,
+  metadata: ProviderMetadata
+): LLMEvent[] {
+  const events: LLMEvent[] = [];
+  const item = event.item;
+
+  switch (item.type) {
+    case "function_call":
+      events.push({
+        type: "tool_call",
+        content: {
+          id: item.call_id,
+          name: item.name,
+          arguments: item.arguments,
+        },
+        metadata,
+      });
+      break;
+    default:
+      break;
+  }
+  return events;
+}
+
+function itemToEvent(
+  item: ResponseOutputItem,
+  metadata: ProviderMetadata
+): LLMEvent | undefined {
+  switch (item.type) {
+    case "message":
+      return {
+        type: "text_generated",
+        content: {
+          text: item.content
+            .map((content) => {
+              if (content.type === "output_text") {
+                return content.text;
+              }
+              return "";
+            })
+            .join("\n"),
+        },
+        metadata,
+      };
+    case "function_call":
+      return {
+        type: "tool_call",
+        content: {
+          id: item.call_id,
+          name: item.name,
+          arguments: item.arguments,
+        },
+        metadata,
+      };
+    case "reasoning":
+      return {
+        type: "reasoning_generated",
+        content: {
+          text: item.summary.map((summary) => summary.text).join("\n"),
+        },
+        metadata,
+      };
+    default:
+      return undefined;
+  }
+}
+
+function responseCompleted(
+  response: Response,
+  metadata: ProviderMetadata
+): LLMEvent[] {
+  const events: LLMEvent[] = response.output
+    .map((item) => itemToEvent(item, metadata))
+    .filter((item): item is LLMEvent => item !== undefined);
+
+  if (response.usage) {
+    events.push({
+      type: "token_usage",
+      content: {
+        inputTokens: response.usage.input_tokens,
```

Comment:
For more readability, you can destructure before:

`const {output_tokens_details, input_tokens_details, total_tokens} = response.usage`